core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,312 @@
+	if forwardToUrl == "" {
+		return AuditLoggerConfig{}, errors.Errorf("No forwardToURL provided")
+	}
+
+	if _, err := models.ParseURL(forwardToUrl); err != nil {
+		return AuditLoggerConfig{}, errors.Errorf("forwardToURL value is not a valid URL")
+	}
+
+	environment := "production"
+	if isDev {
+		environment = "develop"
+	}
+
+	// Split and prepare optional service client headers from env variable
+	headers := []ServiceHeader{}
+	if encodedHeaders != "" {
+		headerLines := strings.Split(encodedHeaders, "\\")
+		for _, header := range headerLines {
+			keyValue := strings.Split(header, "||")
+			if len(keyValue) != 2 {
+				return AuditLoggerConfig{}, errors.Errorf("Invalid headers provided for the audit logger. Value, single pair split on || required, got: %s", keyValue)
+			}
+			headers = append(headers, ServiceHeader{
+				Header: keyValue[0],
+				Value:  keyValue[1],
+			})
+		}
+	}
+
+	return AuditLoggerConfig{
+		ForwardToUrl:   &forwardToUrl,
+		Environment:    &environment,
+		JsonWrapperKey: &jsonWrapperKey,
+		Headers:        headers,
+	}, nil
+}
+
+type AuditLoggerService struct {
+	logger          logger.Logger            // The standard logger configured in the node
+	enabled         bool                     // Whether the audit logger is enabled or not
+	forwardToUrl    string                   // Location we are going to send logs to
+	headers         []ServiceHeader          // Headers to be sent along with logs for identification/authentication
+	jsonWrapperKey  string                   // Wrap audit data as a map under this key if present
+	environmentName string                   // Decorate the environment this is coming from
+	hostname        string                   // The self-reported hostname of the machine
+	localIP         string                   // A non-loopback IP address as reported by the machine
+	loggingClient   HTTPAuditLoggerInterface // Abstract type for sending logs onward
+
+	loggingChannel chan wrappedAuditLog
+	ctx            context.Context
+	cancel         context.CancelFunc
+	chDone         chan struct{}
+}
+
+// Configurable headers to include in POST to log service
+type ServiceHeader struct {
+	Header string
+	Value  string
+}
+
+type wrappedAuditLog struct {
+	eventID EventID
+	data    Data
+}
+
+// NewAuditLogger returns a buffer push system that ingests audit log events and
+// asynchronously pushes them up to an HTTP log service.
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - no configuration")
+	}
```

Comment:
Is this a feature or a quirk? Can we just use the struct type directly instead?
```suggestion
func NewAuditLogger(logger logger.Logger, config AuditLoggerConfig) (AuditLogger, error) {
```

Reply:
AuditLoggerConfig can be nil

Reply:
This is because when configuration fails (which is a normal thing and happens when there is no configuration) the pointer to this struct is nil

Reply:
> when configuration fails (which is a normal thing and happens when there is no configuration) 

IMHO this is needlessly confusing. We should follow standard idiomatic patterns from Go, the application, and the surrounding code. Is there some reason this system needs to break form that I am overlooking?